sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c

```diff
@@ -64,9 +64,27 @@ static bool _rbuf_update_ints(rbuf_t *entry, uint16_t offset, size_t frag_size);
 static rbuf_t *_rbuf_get(const void *src, size_t src_len,
                          const void *dst, size_t dst_len,
                          size_t size, uint16_t tag, unsigned page);
+/* internal add to repeat add when fragments overlapped */
+static int _rbuf_add(gnrc_netif_hdr_t *netif_hdr, gnrc_pktsnip_t *pkt,
+                     size_t offset, unsigned page);
+
+/* status codes for _rbuf_add() */
+enum {
+    RBUF_ADD_SUCCESS,
+    RBUF_ADD_ERROR,
+    RBUF_ADD_REPEAT,
+};
 
 void rbuf_add(gnrc_netif_hdr_t *netif_hdr, gnrc_pktsnip_t *pkt,
               size_t offset, unsigned page)
+{
+    if (_rbuf_add(netif_hdr, pkt, offset, page) == RBUF_ADD_REPEAT) {
```

Comment:
just to be clear: the current version can probably call `rbuf_add` multiple times (more than 2). And with the proposed diff you limit it to max. 2? Shouldn't it be a `while ( ... )` to keep the same functionality?

Reply:
sound about right

Reply:
No, currently `rbuf_add()` is not called more than twice. While syntactically you are correct, it is dependent on context: The recursive call only happens if there are overlapping fragments in the reassembly buffer

https://github.com/RIOT-OS/RIOT/blob/a87c2c9563320776d3b9bf8247c92787c98d6f83/sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c#L113-L124

Since those fragments however are removed in https://github.com/RIOT-OS/RIOT/blob/a87c2c9563320776d3b9bf8247c92787c98d6f83/sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c#L116 (and the 6LoWPAN thread we are currently in is the only one adding those fragments) there is no way, the recursion anchors itself here. So a while loop is unnecessary, since `_rbuf_add()` won't return `RBUF_ADD_REPEAT` in the second call.

Reply:
Okay, thanks for this explanation. A comment in the code might be insightful, then.

Reply:
Mh... Aren't you just biased by seeing the previous version? A newcomer to this will see "ah the function is called twice at most". No explanation needed. A while loop would have required more explaining how it wouldn't result in an endless loop ;)

Reply:
A newcomer will ask **why** the function is called at least 2 times and **what** `RBUF_ADD_REPEAT` really means (;

Reply:
There is already a comment for that (where the function returns) ;-)